### DISCUSSION TITLE
Possible fix to require loop issue

### DISCUSSION THREAD

**Discussion by eladco898** [nodejs/help#4843] [Ideas]
Hi all,
Lets say we have 2 node modules that using each other, e.g. :
**Module A:**
```js
const B = require('./B');
const A = {
};
module.exports = A;
```

**Module B:**
```js
const A = require('./A');
const B = {
};
module.exports = B;
```

So, I did manage to fix it by moving all requires to be after the "module.exports" line, e.g.:

**New Module A:**
```js
const A = {
};
module.exports = A;
const B = require('./B');
```

The motivation was to enable both modules first to exist and then require each other and it indeed fixes the issue.
My question is if there are some disadvantages/known issues that I have missed?

Please note that the problematic services(modules) have a lot of functionality.
Also, the require loop can be indirectly(means loop that involves 3 or more modules) so solutions like breaking it into smaller modules won't help me here

**Comment by aduh95** [comment]
You may find ES modules more convenient for this kind of loop, as the graph resolution and the module execution are done asynchronously, meaning an `import A from './a.js'` statement won't actually block the execution of following statements. More info on the ES modules integration in Node.js can be found at https://nodejs.org/api/esm.html.

**Reply by eladco898** [reply]
OK, thanks.
Anyway, what do you think about my solution?

**Reply by aduh95** [reply]
I'm not very familiar with the CJS module system, but it seems to fit the documented workaround: https://nodejs.org/api/modules.html#modules_cycles
